test/fixtures/rules/std_rule_cases/L036.yml

```
@@ -57,3 +57,47 @@ test_comment_between_select_and_single_select_target:
         -- This is the user's ID.
         FROM
         safe_user
+
+test_multiple_select_targets_some_newlines_missing_1:
+  fail_str: |
+    select
+      a, b, c,
+      d, e, f, g,
+      h
+    from x
+  # The spaces before a, d, and h look odd, but these are places where the
+  # select targets were already on a separate line, and the rule made no
+  # changes.
+  fix_str: |
+    select
+      a,
+    b,
+    c,
+      d,
+    e,
+    f,
+    g,
+      h
+    from x
```

I'm not sure what you consider best practice here, but I'm [most familiar](https://github.com/fishtown-analytics/corp/blob/master/dbt_coding_conventions.md#example-sql) with indenting being used on select targets, and I'd expect the rule to fix this to

```suggestion
    select
        a,
        b,
        c,
        d,
        e,
        f,
        g,
        h
    from x
```

using the `tab_space_size` configuration option. What do you think?

---

I definitely agree, the output above is not the final version of the SQL that users want to see. In fact, if you run `sqlfluff fix` on the file above, the result looks exactly as you suggest. That's because rule `L003` also gets involved, so `L036` is splitting the select targets across lines and `L003` fixes the indentation.

But within these YAML-based automated tests, only one rule runs -- that is, the specific rule being tested. So what we're seeing here is, you might say, intermediate, partially fixed output.

I had noticed this before when working on other rules, and I think it would actually be counterproductive for each rule to try and take on the responsibilities of other rules -- not only would it lead to code bloat, it could lead to unintended looping behavior if the two lint rules disagree on what the resulting code should look like. (Rule A changes something, rule B disagrees and changes it back.)

---

@NiallRees: Regarding the above, I'm open to more discussion and would be happy to make the `tab_space_size` change if you feel strongly about it. I wonder if this sort of question has come up before in the context of other rules. I would expect such things to become more common over time as we add more and more rules with automated fixes.

---

I think there is definitely a broader discussion to be had here around the 'synergy' of rule fixes. It seems the key question here is should rule fixes be allowed to trigger a linting error for another rule? I have a few other thoughts around overlap which we're starting to see in a couple of rules too. I'll make an issue for our next maintainers' call to run through.

---

Sounds good. A couple other thoughts:
* Depending on the surrounding code, the appropriate indentation would vary (1 tab, 2 tabs, etc.)
* There is some handling around rules triggering rules. @alanmcruickshank may remember an early bug I found where rules were triggering one another infinitely, so the "fix" command would hang forever. He made a change to mitigate this -- I think it's a configurable limit on how many passes "fix" will make before stopping.

---

I'm very torn on this:
- I think rules should be useful on their own.
- I think rules should be as focussed onto just what they need to do.
- I think rule fixes should try to avoid causing other issues.

The effect of all that means that any rules which require indentation or reflowing code either have to reimplement `L003` or do it badly and then hope that the other rule picks up the mess. Both aren't great. I'm also not sure about the third option where lots of rules end up importing `L003` just to use that logic because that feels like it could cause lots of bugs. I'd be happy to merge this rule as is for now, but I agree with @NiallRees that this would be a good one to talk through to get some consensus on.

---

I'm torn, too. A few observations:
* SQLFluff has _many_ rules related to whitespace formatting, and I think those will come up often as "supporting" other rules.
* When a rule triggers another rule, IIRC the knock-on rules don't appear in the list of warnings from `sqlfluff lint` or `sqlfluff fix`, so the user generally won't notice unless they've *disabled* the downstream rules.

---

I also have an example where I'm fixing something that leaves a mess for downstream rules to pick up (but the user never sees it)

https://github.com/sqlfluff/sqlfluff/pull/934#discussion_r609036620

As long as the user doesn't see it, I don't think we should work about it too much. And to Barry's point, if they disabled the rules that "clean up" our messes, then I can only assume that those rules weren't important to them in the first place 😜

I totally get wanting each rule to useful on it's own though. My rule being messy was probably more of a technical limitation on my end rather than a limitation of what's possible with `sqlfluff`

---

I suggest we merge this PR and then deal with the interdependency issue in a separate issue. It has implications wider than just this rule.